File: 3A/se382/A2/lsystem/KochCurve.java
/*
 *
 */
package lsystem;

import java.awt.Rectangle;
import java.awt.geom.Point2D;
import java.awt.geom.Point2D.Double;
import java.util.Vector;


public class KochCurve extends AbstractIteratedSystem {

    private String curString = null;
    private TurtleGraphics turtle = new TurtleGraphics();
    private Vector<Point2D.Double> points = new Vector<Point2D.Double>();
    
    public KochCurve() {
        super(7);
        this.reset();
    }
    
    public String getName() {
        return "Koch Curve";
    }
    
    public void reset() {
        super.reset();
        curString = "F";
        turtle.setLocation(new Point2D.Double(0,0));
        turtle.pointRight();
        points.clear();
        points.add(turtle.getLocation());
    }
    
    protected void iterateSystem() {
        // Algorithm source: http://en.wikipedia.org/wiki/L-system

        StringBuffer newString = new StringBuffer();

        this.turtle.setLocation(new Point2D.Double(0,0));
        this.turtle.pointRight();
        this.points.clear();
        this.points.add(this.turtle.getLocation());

        for (int i = 0; i < curString.length(); i++) {
            switch (curString.charAt(i)) {
                case 'F':
                    this.points.add(this.turtle.moveForward());
                    newString.append("F+F-F-F+F");
                    break;
                case '-':
                    this.turtle.turn(90);
                    newString.append(curString.charAt(i));
                    break;
                case '+':
                    this.turtle.turn(-90);
                    newString.append(curString.charAt(i));
                    break;
            }
        }
        this.curString = newString.toString();
    }

    public Vector<Double> getLines(Rectangle displayArea) {
        return this.scalePointsToArea(points, displayArea);
    }
}

File: 3A/se382/A1/lsystem/KochCurve.cpp
#include "KochCurve.h"

#define MAX_ITERATIONS 5

namespace cs349
{

KochCurve::KochCurve() : IteratedSystem(MAX_ITERATIONS)
{
	this->Reset();
}

KochCurve::~KochCurve()
{
}

string KochCurve::GetName() const
{
	return string("Koch Curve");
}

void KochCurve::SetParamValue(int paramNum, int value)
{
	; // no-op
}

void KochCurve::Reset()
{
	IteratedSystem::Reset();
	curString = string("F");
	this->turtle.SetLocation(Point(0,0));
	this->turtle.PointRight();
	this->points.clear();
	this->points.push_back(this->turtle.GetLocation());
}

void KochCurve::IterateSystem()
{
	// Algorithm source: http://en.wikipedia.org/wiki/L-system

	string newString;

	this->turtle.SetLocation(Point(0,0));
	this->turtle.PointRight();
	this->points.clear();
	this->points.push_back(this->turtle.GetLocation());

	for (string::iterator iter=this->curString.begin(); iter != this->curString.end(); iter++) {
		char c = (*iter);
		if (c == 'F') {
			this->points.push_back(this->turtle.MoveForward());
			newString.append("F+F-F-F+F");
		} else if (c == '-') {
			this->turtle.Turn(90);
			newString.push_back(c);
		} else if (c == '+') {
			this->turtle.Turn(-90);
			newString.push_back(c);
		}
	}
	this->curString = newString;
}

vector<Point> KochCurve::GetLines(const Rectangle & displayArea) const
{
	return this->ScalePointsToArea(this->points, displayArea);
}

}
